File: packages/javascript/src/exceptions.ts
/** Raised when ``onDetect: "block"`` mode detects PII in the input. */
export class PIIDetectedError extends Error {
  public readonly entities: import("./result.js").DetectedEntity[];

  constructor(message: string, entities: import("./result.js").DetectedEntity[]) {
    super(message);
    this.name = "PIIDetectedError";
    this.entities = entities;
    // Restore prototype chain (required for extending Error in TS)
    Object.setPrototypeOf(this, PIIDetectedError.prototype);
  }
}

/** Raised when an optional integration package is not installed. */
export class MissingDependencyError extends Error {
  constructor(pkg: string, extra: string) {
    super(
      `"${pkg}" is not installed. Run: npm install ${pkg}\n` +
        `(required for prompt-sanitizer ${extra} support)`
    );
    this.name = "MissingDependencyError";
    Object.setPrototypeOf(this, MissingDependencyError.prototype);
  }
}

/**
 * Raised when a replacement token already maps to a *different* original
 * value. This should only happen if a Vault was hydrated from a persisted
 * snapshot without correctly restoring/reconciling its counters — it is
 * a loud failure by design, since silently overwriting the mapping would
 * make an old placeholder deanonymize to the wrong value.
 */
export class VaultCollisionError extends Error {
  public readonly replacement: string;
  public readonly existingOriginal: string;
  public readonly incomingOriginal: string;

  constructor(replacement: string, existingOriginal: string, incomingOriginal: string) {
    super(
      `Replacement token "${replacement}" is already mapped to a different ` +
        `original value. This usually means a Vault's counters were not ` +
        `restored correctly from a persisted snapshot.`
    );
    this.name = "VaultCollisionError";
    this.replacement = replacement;
    this.existingOriginal = existingOriginal;
    this.incomingOriginal = incomingOriginal;
    Object.setPrototypeOf(this, VaultCollisionError.prototype);
  }
}

/** Raised when a VaultStore load/save/delete fails, or a snapshot's version is unsupported. */
export class VaultStoreError extends Error {
  constructor(message: string, options?: { cause?: unknown }) {
    super(message, options);
    this.name = "VaultStoreError";
    Object.setPrototypeOf(this, VaultStoreError.prototype);
  }
}

File: packages/python/prompt_sanitizer/exceptions.py
from __future__ import annotations

from .entities import EntityType
from .result import DetectedEntity


class PIIDetectedError(Exception):
    """Raised when PII is detected and the sanitizer is configured with on_detect='block'."""

    def __init__(self, entities: list[DetectedEntity]) -> None:
        self.entities = entities
        types = sorted({e.entity_type.value for e in entities})
        count = len(entities)
        super().__init__(
            f"PII detected (block mode): {count} instance(s) of [{', '.join(types)}]"
        )


class MissingDependencyError(ImportError):
    """Raised when an optional dependency required for the chosen Mode is not installed."""

    def __init__(self, package: str, extra: str) -> None:
        super().__init__(
            f"Package '{package}' is required but not installed. "
            f"Install it with: pip install ai-prompt-sanitizer[{extra}]"
        )


class VaultCollisionError(Exception):
    """
    Raised when a replacement token already maps to a *different* original
    value inside a :class:`Vault`.

    This should only happen if a vault was hydrated from a persisted
    snapshot without correctly restoring/reconciling its counters. It is
    a loud failure by design: silently overwriting the mapping would make
    an old placeholder deanonymize to the wrong value.
    """

    def __init__(self, replacement: str, existing_original: str, incoming_original: str) -> None:
        self.replacement = replacement
        self.existing_original = existing_original
        self.incoming_original = incoming_original
        super().__init__(
            f"Replacement token '{replacement}' is already mapped to a different "
            f"original value. This usually means a Vault's counters were not "
            f"restored correctly from a persisted snapshot."
        )


class VaultStoreError(Exception):
    """Raised when a VaultStore load/save/delete fails, or a snapshot's version is unsupported."""
